modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZoneManager.java

```diff
@@ -480,6 +625,152 @@ public int getZoneId(String name) {
         }
     }
 
+    /**
+     * The method for obtaining data nodes of the specified zone. If the dataNodesAutoAdjustScaleUp or dataNodesAutoAdjustScaleDown
+     * equals to 0 then returned completable future will be completed when data nodes will be saved to the meta storage.
+     *
+     * @param zoneId Zone id.
+     * @param topVer Topology version.
+     * @return The data nodes future.
+     */
+    public CompletableFuture<Set<String>> getDataNodes(int zoneId, long topVer) {
+        boolean immediateScaleUp0 = false;
+        boolean immediateScaleDown0 = false;
+
+        if (zoneId == DEFAULT_ZONE_ID) {
+            immediateScaleUp0 = zonesConfiguration.defaultDistributionZone().dataNodesAutoAdjustScaleUp().value() == 0;
+            immediateScaleDown0 = zonesConfiguration.defaultDistributionZone().dataNodesAutoAdjustScaleDown().value() == 0;
+        } else {
+            NamedConfigurationTree<DistributionZoneConfiguration, DistributionZoneView, DistributionZoneChange> zones =
+                    zonesConfiguration.distributionZones();
+
+            for (int i = 0; i < zones.value().size(); i++) {
+                DistributionZoneView zone = zones.value().get(i);
+
+                if (zone.zoneId() == zoneId) {
+                    immediateScaleUp0 = zone.dataNodesAutoAdjustScaleUp() == 0;
+                    immediateScaleDown0 = zone.dataNodesAutoAdjustScaleDown() == 0;
+
+                    break;
+                }
+            }
+        }
+
+        boolean immediateScaleUp = immediateScaleUp0;
+        boolean immediateScaleDown = immediateScaleDown0;
+
+        if (!immediateScaleUp && !immediateScaleDown) {
+
+            DataNodes dataNodes = this.dataNodes.get(zoneId);
+
+            Set<String> nodes;
+
+            if (dataNodes == null) {
+                nodes = emptySet();
+            } else {
+                nodes = this.dataNodes.get(zoneId).nodes();
+            }
+
+            return completedFuture(nodes);
+        }
+
+        CompletableFuture<Set<String>> dataNodesFut = new CompletableFuture<>();
+
+        CompletableFuture<Void> topVerFut;
+
+        synchronized (dataNodesMutex) {
+            if (topVer > lastTopVer) {
+                topVerFut = topVerFutures.get(topVer);
+
+                if (topVerFut == null) {
+                    topVerFut = new CompletableFuture<>();
+
+                    topVerFutures.put(topVer, topVerFut);
```

Comment:
Could we use `putIfAbsent` here instead all of checks for null?

Reply:
Thanks, I go with `computeIfAbsent`